src/core/lib/security/credentials/oauth2/oauth2_credentials.cc

```diff
@@ -472,6 +474,140 @@ grpc_call_credentials* grpc_google_refresh_token_credentials_create(
       .release();
 }
 
+//
+// STS credentials.
+//
+
+namespace {
+
+char* maybe_add_to_query(char* body, const char* field_name, char* field) {
+  if (field == nullptr || strlen(field) == 0) return body;
+  char* new_body;
+  gpr_asprintf(&new_body, "%s&%s=%s", body, field_name, field);
+  gpr_free(body);
+  return new_body;
+}
+
+class grpc_sts_token_fetcher_credentials
+    : public grpc_oauth2_token_fetcher_credentials {
+ public:
+  grpc_sts_token_fetcher_credentials(
+      grpc_uri* sts_url,  // Ownership transfered.
+      const grpc_sts_credentials_options* options)
+      : sts_url_(sts_url),
+        resource_(gpr_strdup(options->resource)),
+        audience_(gpr_strdup(options->audience)),
+        scope_(gpr_strdup(options->scope)),
+        requested_token_type_(gpr_strdup(options->requested_token_type)),
+        subject_token_(gpr_strdup(options->subject_token)),
+        subject_token_type_(gpr_strdup(options->subject_token_type)),
+        actor_token_(gpr_strdup(options->actor_token)),
+        actor_token_type_(gpr_strdup(options->actor_token_type)) {}
+
+  ~grpc_sts_token_fetcher_credentials() override { grpc_uri_destroy(sts_url_); }
+
+ protected:
+  void fetch_oauth2(grpc_credentials_metadata_request* metadata_req,
+                    grpc_httpcli_context* http_context,
+                    grpc_polling_entity* pollent,
+                    grpc_iomgr_cb_func response_cb,
+                    grpc_millis deadline) override {
+    grpc_http_header header = {(char*)"Content-Type",
+                               (char*)"application/x-www-form-urlencoded"};
+    grpc_httpcli_request request;
+    char* body = nullptr;
+    gpr_asprintf(&body, GRPC_STS_POST_MINIMAL_BODY_FORMAT_STRING,
+                 subject_token_.get(), subject_token_type_.get());
+    body = maybe_add_to_query(body, "resource", resource_.get());
+    body = maybe_add_to_query(body, "audience", audience_.get());
+    body = maybe_add_to_query(body, "scope", scope_.get());
+    body = maybe_add_to_query(body, "requested_token_type",
+                              requested_token_type_.get());
+    body = maybe_add_to_query(body, "actor_token", actor_token_.get());
+    body =
+        maybe_add_to_query(body, "actor_token_type", actor_token_type_.get());
+
+    memset(&request, 0, sizeof(grpc_httpcli_request));
+    request.host = (char*)sts_url_->authority;
+    request.http.path = (char*)sts_url_->path;
+    request.http.hdr_count = 1;
+    request.http.hdrs = &header;
+    request.handshaker = (strcmp(sts_url_->scheme, "https") == 0)
+                             ? &grpc_httpcli_ssl
+                             : &grpc_httpcli_plaintext;
+    /* TODO(ctiller): Carry the resource_quota in ctx and share it with the host
```

Comment:
maybe assign TODO to another person.

Reply:
Maybe in another PR as there are a bunch of these on the same pattern. What do you think?

Reply:
SG